pkg/cache/sql/store/store.go

```diff
@@ -0,0 +1,370 @@
+/*
+Package store contains the sql backed store. It persists objects to a sqlite database.
+*/
+package store
+
+import (
+	"bytes"
+	"context"
+	"database/sql"
+	"encoding/gob"
+	"fmt"
+	"reflect"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/rancher/lasso/pkg/cache/sql/db"
+	"github.com/rancher/lasso/pkg/cache/sql/db/transaction"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	_ "modernc.org/sqlite"
+)
+
+const (
+	upsertStmtFmt   = `REPLACE INTO "%s"(key, object, objectnonce, dek, deknonce) VALUES (?, ?, ?, ?, ?)`
+	deleteStmtFmt   = `DELETE FROM "%s" WHERE key = ?`
+	getStmtFmt      = `SELECT object, objectnonce, dek, deknonce FROM "%s" WHERE key = ?`
+	listStmtFmt     = `SELECT object, objectnonce, dek, deknonce FROM "%s"`
+	listKeysStmtFmt = `SELECT key FROM "%s"`
+)
+
+// Store is a SQLite-backed cache.Store
+type Store struct {
+	c       cache.Indexer
+	name    string
+	encrypt bool
+	typ     reflect.Type
+	keyFunc cache.KeyFunc
+	dbLock  *sync.RWMutex
+
+	DBClient
+
+	shouldEncrypt bool
+
+	upsertStmt   *sql.Stmt
+	deleteStmt   *sql.Stmt
+	getStmt      *sql.Stmt
+	listStmt     *sql.Stmt
+	listKeysStmt *sql.Stmt
+
+	afterUpsert []func(key string, obj any, tx db.TXClient) error
+	afterDelete []func(key string, tx db.TXClient) error
+}
+
+// Test that Store implements cache.Indexer
+var _ cache.Store = (*Store)(nil)
+
+type DBClient interface {
+	Begin() (db.TXClient, error)
+	Prepare(stmt string) *sql.Stmt
+	QueryForRows(ctx context.Context, stmt transaction.Stmt, params ...any) (*sql.Rows, error)
+	ReadObjects(rows db.Rows, typ reflect.Type, shouldDecrypt bool) ([]any, error)
+	ReadStrings(rows db.Rows) ([]string, error)
+	Upsert(tx db.TXClient, stmt *sql.Stmt, key string, obj any, shouldEncrypt bool) error
+	CloseStmt(closable db.Closable) error
+	// TxModeClient() DBClient
+}
+
+var backoffRetry = wait.Backoff{Duration: 50 * time.Millisecond, Factor: 2, Steps: 10}
+
+// NewStore creates a SQLite-backed cache.Store for objects of the given example type
+func NewStore(example any, keyFunc cache.KeyFunc, c DBClient, shouldEncrypt bool, name string) (*Store, error) {
+	s := &Store{
```

Comment:
Some workflows result in `panic`, so please add `nil` checks for `keyFunc` and `c`. We probably should add a `nil` check for the return of `reflect.TypeOf(example)` as well.

Reply:
To me if either `keyFunc` or `c` are nil then it is a programmer error without any reasonable ways to recover, so I am fine panicking.

Let's have a discussion via chat if you have strong perspectives on this one.